2024-11 update of Alokai Next.js Guide (#7313)
vuestorefront/vue-storefront: docs/content/guides/2.alokai-essentials/1.alokai-next-js/9.udl.md

#### transformImageUrl

> SAP Commerce Cloud stores image urls as relative paths but the storefront needs absolute paths to display images. `transformImageUrl` function is responsible for providing correct image url.

Reviewer: A quick explanation of what `transformImageUrl` does would help. Does it add the base path? How is this base path configured? Just a one sentence of what it does to eliminate any confusion.

### Configuring Alokai SDK

> To configure Alokai SDK with the Unified Data Layer, we need to change the SDK configuration. Open `storefront/sdk/sdk.ts` file and add the following code:

Reviewer: ```suggestion
To configure Alokai SDK with the Unified Data Layer, we need to change the SDK configuration. In `storefront/sdk/sdk.ts`, add the following code that uses the `UnifiedEndpoints`:
```

    </section>
  );
}
```

> Now we need to pass a proper data to `ProductDetails` component. Apply the following changes to `product/[id]/page.tsx` file:

Reviewer: ```suggestion
Now, we need to pass the proper data to `ProductDetails` component. Apply the following changes to `product/[id]/page.tsx` file:
```

```diff
export default async function Page({ params }: { params: { id: string } }) {
  const sdk = getSdk();

-  const { data } = await sdk.sapcc.getProduct({
-    productCode: params.id,
-  });

+  const { product } = await sdk.unified.getProductDetails({
+    id: params.id,
+  });

  return (
    <div className="flex flex-col gap-8 md:gap-12 lg:gap-16 max-w-screen-xl m-auto px-4 md:px-8 lg:px-12 xl:px-16 py-8 md:py-12 lg:py-16 xl:py-20">
      <section className="flex flex-col items-start gap-8 md:flex-row md:gap-4 xl:gap-6">
        <ProductGallery />
-        <ProductDetails product={data} />
+        <ProductDetails product={product} />
      </section>
      <ProductSlider />
    </div>
  );
}
```

Finally, if you run the application, you will see that both home page and product details page are working as expected. Except for one thing. If you try to add a product to the cart, you will see that we have an issue with the `useCart` hook - nothing happens when you click the "Add to cart" button.

> I will leave this as an exercise for you to fix. You can use the approach we learned above to fix the `useCart` hook. Only, I will give you a small hint - you need to replace the `addToCart` method from the `sapcc` module with the `addToCart` method from the `unified` module and as well, you need to replace the `Cart` type with the `SfCart` type from the Unified Data Model. You can also find the solution here [udl branch](https://github.com/vuestorefront-community/nextjs-starter/tree/udl).

Reviewer: ```suggestion
As an additional exercise, you can use the approach we learned above to fix the `useCart` hook. As a starting point, you'll need to replace the `addToCart` method from the `sapcc` module with the `addToCart` method from the `unified` module and as well, you need to replace the `Cart` type with the `SfCart` type from the Unified Data Model. You can also find the solution here [udl branch](https://github.com/vuestorefront-community/nextjs-starter/tree/udl).
```

You can find more information about different Unified Api Methods in the [Unified Cart Methods](https://docs.vuestorefront.io/storefront/unified-data-layer/unified-methods/cart) section of Storefront documentation.

## Displaying Images

> So far we've been displaying hardcoded images. Now thanks to the Unified Data Layer we can easily access images from the e-commerce platform.

Reviewer: ```suggestion
So far, we've been displaying hardcoded images. But with the Unified Data Layer, we can easily access images from the e-commerce platform.
```

```tsx
// ... rest of the code
import { Product } from '@vsf-enterprise/sap-commerce-webservices-sdk';

interface ProductGalleryProps {
  images: SfProduct['gallery'];
}
// ... rest of the code
```

In the above code, we have created a TypeScript interface `ProductGalleryProps` for the props of the `ProductGallery` Block. We have used the `SfProduct` type from the UDL.

Now, we will use the `ProductGalleryProps` interface to define the type of the `props` of the `ProductGallery` component. Replace the content of the `storefront/components/ProductGallery.tsx` file with the following code:

```diff
- export default function GalleryVertical() {
+ export default function GalleryVertical({ images }: ProductGalleryProps) {
```

We can also remove the `images` constant and `withBase` function from the `ProductGallery` component as we are now passing the `images` prop from the parent component.

Now, Replace all the occurrences of `imageThumbSrc` and `imageSrc` with `url`. The final shape of the `ProductGallery` component can be found in the [udl branch](https://github.com/vuestorefront-community/nextjs-starter/tree/udl).

:::info

> Your IDE might yell that we're using `<img>` instead of `next/image`. That's because we've copied the Gallery component from Storefront UI  which is agnostic of a meta framework.

Reviewer: ```suggestion
Your IDE might throw a warning that we're using `<img>` instead of `next/image`. That's because we've copied the Gallery component from Storefront UI  which is agnostic of a meta framework.
```
